[MyQ] Initial commit of the MyQ binding for OH3

digitaldan opened the pull request:
Signed-off-by: Dan Cunningham <dan@digitaldan.com>

This introduces a new MyQ binding for openHAB to replace the 1.x version.   Feedback welcome. 

This will replace the 1.x myq binding referenced in #6179

morph166955 commented:
I've tested the binary you provided.  I'm getting errors in the log showing:

```
2020-12-13 13:35:05.518 [INFO ] [me.event.ThingStatusInfoChangedEvent] - Thing 'myq:account:home' changed from UNKNOWN to ONLINE
2020-12-13 13:35:05.615 [INFO ] [me.event.ThingStatusInfoChangedEvent] - Thing 'myq:account:home' changed from ONLINE to OFFLINE (COMMUNICATION_ERROR): Invalid Response Code 404

```

digitaldan commented:
Are you configuring this with things files?  Auto discovery ?  Can you describe the steps you took to configure this?   Also it seems the jar you have might be a few revisions behind (logging looks outdated).  Can you download again from https://github.com/digitaldan/openhab-addons/releases/tag/myq-wip-1 ?  I update that with the last build time and a new jar instead of creating new revisions.  Any other logging or info would be helpful. Thanks!

digitaldan commented:
Actually the logging does not look outdated, my guess is you do have a updated version, so info on how you configured this would be helpful, thanks!

morph166955 commented:
Configured as:

```
Bridge myq:account:home "MyQ Account" [ username="my@mail.com",password="shhhhhh",refreshInterval=60 ] {
        Thing garagedoor ABCDE "Double Garage Door" [ serialNumber="ABCDE" ]
        Thing garagedoor FGHIJ "Single Garage Door" [ serialNumber="FGHIJ" ]
}

```

digitaldan commented:
can you download org.openhab.binding.myq-3.0.0-SNAPSHOT.jar and try again?  For some reason i don't think it liked the `UserId` i was grabbing for the account (although it works for me) , so i 'm grabbing a different one .

digitaldan commented:
sorry , download at https://github.com/digitaldan/openhab-addons/releases/download/myq-wip-1/org.openhab.binding.myq-3.0.0-SNAPSHOT.jar

morph166955 commented:
I'd say closer but still missing something.  The JSON that starts with Users is now about twice as long as it was with the last version.  I still get the 404.401 error however after that.

digitaldan commented:
Try it again ?  I  missed a spot where i needed to use that new ID , thanks.

morph166955 commented:
There we go!  I'll start to test/monitor actual usage now.

digitaldan commented:
Thanks for testing, appreciate it !

digitaldan commented:
This is now ready for review, thanks!

morph166955 commented:
Do you have a final JAR you want me to test out to confirm it's all solid here?

digitaldan commented:
Sure, i just updated:
https://github.com/digitaldan/openhab-addons/releases/download/myq-wip-1/org.openhab.binding.myq-3.0.0-SNAPSHOT.jar

digitaldan commented:
@computergeek1507  thanks for helping with the light. I'm not sure how you made those changes  as i don't see a branch they were committed from or a sign off message.  In fact , some how you modified my personal branch on my repo.   If we change that UID to lamp, we also need to modify the thing definition to match, i will update that myself, and make sure things compile and work as much as i can..

digitaldan commented:
@computergeek1507 is it possible to remove your changes?  I'm not sure now how to get my branch in sync with those changes.

fwolter commented on bundles/org.openhab.binding.myq/pom.xml:
```diff
@@ -0,0 +1,17 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.0.0-SNAPSHOT</version>
```
To make the binding compile, you need to rebase your branch and set the version to the upcoming openHAB version: `3.1.0-SNAPSHOT`.

Here are the commands for rebasing your branch:

If not already done, add the upstream openHAB addon repo as a remote to your local repo and fetch it:
```
git remote add upstream https://github.com/openhab/openhab-addons.git
git fetch upstream
```
Then, you can rebase your PR's branch onto `main`:
```
git rebase upstream/main
```
Finally force-push the rebased branch to this PR's branch:
```
git push origin [your branch name of this PR] --force-with-lease
```

digitaldan commented:
@computergeek1507 I updated the binding to use "lamp" vs "light" throughout the binding so there is consistency with the MyQ naming conventions.  I updated the jar at https://github.com/digitaldan/openhab-addons/releases/download/myq-wip-1/org.openhab.binding.myq-3.0.0-SNAPSHOT.jar  with these changes.  If you previously added a lamp/light, you will need to remove it and let the binding find the lamp again.  Thanks for the pointer.  If you have a json dump of a lamp that would be helpful. 

@fwolter i will update it to the latest milestone shortly, thanks.

computergeek1507 commented:
@digitaldan I sent the JSON to you over on the community forum

digitaldan commented:
I totally missed that , thank you!  Looking at the json and my recent changes, it seems like it should work.  Let me know if thats the case or not.

digitaldan commented:
@fwolter, done!

digitaldan commented:
I have updated the binding from feedback from @computergeek1507  (Thanks!).  This is ready for review !

morph166955 commented:
Just for reference since I tagged this PR on an issue, I don't believe (right now) that this binding is causing a problem with that issue.  I was noting for completeness that this was the only binding in my addons folder right now.  

To note, I've compiled a jar of this with your latest commits, it's running without issue.

digitaldan commented:
ping @fwolter  @cpmeister , i'm hoping to get this 1.x replacement in for 3.1.0 :-)

digitaldan commented:
>i'm hoping to get this 1.x replacement in for 3.1.0 :-)

oof, I meant for 3.0.1 (or 3.0.2) , 3.1 could be awhile :-)

fwolter commented:
Hm, the next milestone here is 3.1 :o

digitaldan commented:
>i'm hoping to get this 1.x replacement in for 3.1.0 :-)

Just ignore me, i have no idea what i'm talking about  🤦

digitaldan commented:
👀

fwolter commented on bundles/org.openhab.binding.myq/README.md:
```diff
@@ -0,0 +1,71 @@
+
+This represents the MyQ cloud account and uses the same credentials needed when using the MyQ mobile application.
+
+ThingTypeUID:`account`
```
Same for below.
```suggestion
ThingTypeUID: `account`
```

fwolter commented on bundles/org.openhab.binding.myq/README.md:
```diff
@@ -0,0 +1,71 @@
+
+### Garage Door
+
+This represents a garage door associated with an account.  Multiple garage doors are supported.
```
Same for below.
```suggestion
This represents a garage door associated with an account. Multiple garage doors are supported.
```

fwolter commented on bundles/org.openhab.binding.myq/README.md:
```diff
@@ -0,0 +1,71 @@
+| Channel       | Item Type     | Thing Type       | Description              |
+|---------------|---------------|------------------|--------------------------|
+| status        | String        | garagedoor, lamp | Status of the device     |
+| switch        | Switch        | garagedoor, lamp | ON (open), OFF (close)   |
+| contact       | Contact       | garagedoor       | Open, Closed             |
+| rollershutter | Rollershutter | garagedoor       | UP (open), DOWN (closed) |
```
Is there any reason why these are two separate channels? What's the difference to the `switch` channel?

digitaldan replied:
Some users prefer this to be a switch, similar to how locks are represented, others consider the door more like a rollershutter.  I tried to support both use cases.

Hilbrand replied:
There should be only 1 channel. Contact conceptually is readonly. Rollesshutter seems the most logically,but should accept percentage and move/stop. So that should be handled. See also https://www.openhab.org/docs/configuration/items.html#type

digitaldan replied:
thanks @Hilbrand for the feedback
>Rollesshutter seems the most logically,but should accept percentage and move/stop.

I would certainly like to , but since the api only responds to open/close and not a % or stop, and does not report %, how would you suggest i do that? 

>There should be only 1 channel.

Ideally that would be awesome, but its not very black and white what item type is in use here. From a UI perspective, a rollershutter makes the most sense,  our widgets specifically have a UP/Down interface, but as I mentioned above, we can not support % or stop.  From a contact point of view this also makes sense, and in fact is useful when using this in a group of contacts with sensors  for perimeter monitoring (which i do) , but of course there is not a command type for a contact, so that can not be the only channel.  Many users have used a simple switch to control this.  I am trying to appease the current users of previous myq bindings that have been floating around.

Is there another way to map multiple item types to a single channel?  In this case, i think the types i have here all make very logical sense, but picking only one will greatly limit the usability to our existing myq users. 

i am also taking a queue from other bindings,  like our Open Garage addon, which handles this in a similar way.   

I am open to suggestions, but i'm hoping this is not going to delay getting this 1.x replacement merged in.

digitaldan replied:
Regarding the rollershutter, you are correct i was not handling % commands and should have, so i have added that in my last commit.

fwolter replied:
I go for the Contact type too, but what is the use case for the Switch in addition to the Rollershutter?

digitaldan replied:
I can live without the switch, it seemed to me a common UI element that users of the previous bindings were using, but a rollershutter covers this use case nicely.  I personally don't use it, so I'll remove it and see if anyone screams.

computergeek1507 replied:
I use the switch, please don't remove it. Other bindings have dimmers and on/offs for the same thing. I don't see how less choice is good thing.

digitaldan replied:
I'm ok leaving it in , although dimmers and switches are a different comparison, dimmer items also support ON/OFF commands and states, so you can map a Switch Item  to a dimmer channel and it will work, so i don't know a lot of bindings that do both channel types for the same function, or at least they should not.

morph166955 replied:
I also use the switch.  It just makes some things cleaner across my rules.

digitaldan replied:
I left the switch there in my last commit, i vote we keep it for consistency with existing user's expectations.

ccutrer replied:
Just putting in my two cents - I’ve used a myQ plugin of some form for a few years now. When I first came to it, I was extremely confused that their were so many channels that seemingly did the same thing. Especially since the Rollershutter channel didn’t respond to 0/100 which is what some things (HomeKit) automatically do instead of UP/DOWN. I saw you fixed that last part recently, and that’s nice. There’s just something to be said for presenting the simplest possible interface (with full functionality), and making users adapt it via rules if they want to use it say in a group with slightly different item types. As long as documentation is clear that these channels all actually do the same thing, I won’t argue against keeping them though.

fwolter replied:
IMHO the current number of Channels for the same function are too much and therefore confusing. As several users seem to use the Switch, my suggestion would be to keep this and remove all others. As the garage door cannot be controlled with 0-100%, but only open/close, the Rollershutter doesn't have any benefit over the Switch. Keep it simple... If the user wants to do some more complex things, it's still possible by using rules. I think this fits good with openHAB concepts.

morph166955 replied:
I would agree with getting rid of the rollershutter and the contact.  Personally I use the switch and the status channels.  The switch to send commands, and the status for my rules (so that I can identify open vs. opening vs. closing. vs. closed).  If you really want it to say up/down, you can easily do that in the sitemap.

digitaldan replied:
So i feel very strongly about keeping a roller shutter.  I never once refer to my door as being ON or OFF.   In our UIs the rollershutter widget makes sense, an on/off switch does not and is confusing IMHO (why is a door that is open ON? ).    

>cannot be controlled with 0-100%

The binding does support 0%/100% as well as UP/DOWN, i don't think is uncommon for simple  roller shutters to  work this way, i have a unit which only goes up or down, not a %, i believe the motor simply inverts direction when the current is inverted. 

 < If the user wants to do some more complex things, it's still possible by using rules. 

If we are already working around an issue like this with rules before we have even submitted the binding i think we have lost from a usability POV.  

I would like  drop the contact channel and keep the others as it is, i do feel strongly about this.  My reasoning is there a small amount of channels in the binding,  i don't think it's going to confuse users, there are other bindings that do this, and  the fact that we have so many opinions on this i think proves the value of keeping both channels types .

morph166955 replied:
I'm cool with that.

fwolter commented on bundles/org.openhab.binding.myq/README.md:
```diff
@@ -0,0 +1,71 @@
+
+| Channel       | Item Type     | Thing Type       | Description              |
+|---------------|---------------|------------------|--------------------------|
+| status        | String        | garagedoor, lamp | Status of the device     |
```
Can you mention which state the user can expect?

digitaldan replied:
There are state descriptions in the XML as well.

fwolter replied:
Yeah, but do you want to ask the user to read the source? Better document it here.

digitaldan replied:
fair enough

fwolter commented on bundles/org.openhab.binding.myq/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+	Copyright (c) 2010-2020 Contributors to the openHAB project
+
+	See the NOTICE file(s) distributed with this work for additional
+	information.
+
+	This program and the accompanying materials are made available under the
+	terms of the Eclipse Public License 2.0 which is available at
+	http://www.eclipse.org/legal/epl-2.0
+
+	SPDX-License-Identifier: EPL-2.0
+
+-->
```
The features.xml doesn't need a license header.

fwolter commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyQHandlerFactory.java:
```diff
@@ -0,0 +1,95 @@
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (THING_TYPE_ACCOUNT.equals(thingTypeUID)) {
+            MyQDiscoveryService discoveryService = new MyQDiscoveryService();
+            MyQAccountHandler accountHandler = new MyQAccountHandler((Bridge) thing, httpClient, discoveryService);
+            discoveryServiceRegs.put(accountHandler.getThing().getUID(), bundleContext
+                    .registerService(DiscoveryService.class.getName(), discoveryService, new Hashtable<>()));
```
You could use `ThingHandlerService` which would make this code unnecessary. The documentation for this is not yet merged, but available in the PR: https://github.com/openhab/openhab-docs/pull/1262/files#diff-c4a4d8725430bc2ea046182bfc73ac51349d989c7f6a8a6fa3001a226a09ad98R932

digitaldan replied:
Thanks, but my discovery service does not need a reference to the bridge, rather the reverse.

fwolter replied:
Why not make it vice versa via a callback? Then, you could use the ThingHandlerService, which would make the code more maintainable.

digitaldan replied:
I looked at that, but it seemed to complicate things without any clear benefit. Right now the handlerFactory knows about the thingHandler and discoveryService and  the thingHandler also knows about the discoveryService, if i use a callback, then now the discoveryService has to know about the thingHandler as well, we now have all 3 of the classes knowing about each other which seems to be a circular dependency without a clear benefit.

fwolter commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -0,0 +1,330 @@
+    private @Nullable String securityToken;
+    private @Nullable AccountDTO account;
+    private MyQDiscoveryService discoveryService;
+    private Integer normalRefreshInterval = 60;
+    private Integer rapidRefreshInterval = 5;
```
Primitive types should be used where possible.

It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

Can these be static final?
```suggestion
    private int normalRefreshInterval = 60;
    private int rapidRefreshInterval = 5;
```

fwolter commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -0,0 +1,330 @@
+                        }
+                        // did not find a thing, add it to the inbox
+                        ThingUID thingUID = new ThingUID(thingTypeUID, getThing().getUID(), id);
+                        discoveryService.deviceDiscovered(thingUID, getThing().getUID(), device.name);
```
When using the ThingHandlerService, you could set a reference to the discovery service in `ThingHandlerService.setThingHandler()` to make it acccessible here.

fwolter commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -0,0 +1,330 @@
+            logger.trace("Account Response - status: {} content: {}", result.responseCode, result.content);
+            return result;
+        } catch (ExecutionException e) {
+            logger.debug("request failed", e);
```
Are you logging the stack trace by intention?

fwolter reviewed (changes requested):
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

digitaldan commented:
Thanks @fwolter !  I have incorporated your feedback into my last commit.

digitaldan commented:
👀

fwolter commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -0,0 +1,329 @@
+        }
+    }
+
+    private void getDevices() throws InterruptedException {
```
If I see correctly this method does the (background) discovery of devices. So, it's better to move this into the discovery service. Then, you won't have a problem with the depdendencies of handler and discovery service either.

digitaldan replied:
So, when i poll the API i get all devices, so every poll cycle (60 seconds default ) i have the opportunity to discover new devices. I assumed this would be more efficient than having to schedule a separate call to the bridge to do an additional poll just to discover devices, that seemed less efficient with more complexity.

fwolter replied:
Maybe the API has the ability to query only the state of one devices at a time? If not, I would do a discovery only if the user triggered a scan. That would reduce overall code complexity.

digitaldan replied:
Its' a single api call that i'm using, the API is not documented and heavily rate limited, its not clear even if there is a single device endpoint (which i have not seen), if you had more then one, how that would blow through the rate limiting polling for each.  FYI i have a 2 bay garage, so i have 2 units under the same account.  

>That would reduce overall code complexity.

I'm all for reducing complexity, but i'm not sure i see that happening, we would have the same discovery code that exists, PLUS i would need to add additional methods to call the same poll code, but this time not to update devices, just for discovery,  its going to be more code, additional coordination of outgoing requests and much more complex IMHO.

fwolter replied:
It's reasonable to make one request for all devices due to the rate limit. Nevertheless, the discovery service could make its own request as it is not triggered permanently. This way you could cut off the counter intuitive reverse dependency of bridge handler and discovery service and you can implement the ThingHandlerService. Of course you could make `getDevices()` stateless and reuseable to use it in the bridge handler and the discovery service.

digitaldan replied:
Hi @fwolter  i have refactored the discover class to use the ThingHandlerService .  I did not realize this removed the need for all the boiler plate code in the handler factory to create, track and destroy the discovery service, so i agree this was worth the refactoring.

fwolter replied:
Great!

Hilbrand commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -107,6 +109,24 @@ public void dispose() {
+        DevicesDTO localDeviceCaches = devicesCache;
+        if (localDeviceCaches != null) {
+            DeviceDTO device = localDeviceCaches.items.stream()
+                    .filter(d -> childHandler.getThing().getUID().getId().equalsIgnoreCase(d.serialNumber)).findFirst()
```
You're using the id as a unique identifier.  However the convention is that the user can give it any name allowed. So it should not be used as identifier. You should check here on the property `serialNumber`.

digitaldan replied:
Good catch !

Hilbrand commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -267,10 +269,11 @@ private void getDevices() throws InterruptedException {
                         for (Thing thing : getThing().getThings()) {
                             // support both upper and lower case serialIds
-                            if (thing.getUID().getId().toLowerCase().equals(id)) {
+                            if (thing.getConfiguration().as(MyQDeviceConfiguration.class).serialNumber.toLowerCase()
+                                    .equals(deviceSerial)) {
```
In general it's not recommend to get the configuration everytime. I would suggest to make a method `getSerialNumber()` on the handler and also use `equalsIgnoreCase`. Something like:
```suggestion
                            if (((MyQDeviceHandler) thing.getHandler()).getSerialNumber().equalsIgnoreCase(deviceSerial)) {
```

digitaldan replied:
>In general it's not recommend to get the configuration everytime. 

Yep agreed

>on the handler and also use equalsIgnoreCase

yeah, i'm not sure what i was thinking there

digitaldan replied:
done!

digitaldan commented:
I Just rebased to main and updated license headers to get the build passing

digitaldan commented:
👀

fwolter commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -0,0 +1,352 @@
+    @Override
+    public void childHandlerInitialized(ThingHandler childHandler, Thing childThing) {
+        DevicesDTO localDeviceCaches = devicesCache;
+        if (localDeviceCaches != null) {
+            DeviceDTO device = localDeviceCaches.items.stream()
+                    .filter(d -> ((MyQDeviceHandler) childHandler).getSerialNumber().equalsIgnoreCase(d.serialNumber))
+                    .findFirst().orElse(null);
+            if (device != null) {
+                ((MyQDeviceHandler) childHandler).handleDeviceUpdate(device);
+            }
```
The child handler can be of different type if the user made a mistake in the textual configuration. And syntactical sugar:
```suggestion
        if (localDeviceCaches != null && childHandler instanceof MyQDeviceHandler) {
            MyQDeviceHandler handler = (MyQDeviceHandler) childHandler;
            localDeviceCaches.items.stream()
                    .filter(d -> ((MyQDeviceHandler) childHandler).getSerialNumber().equalsIgnoreCase(d.serialNumber))
                    .findFirst().ifPresent(handler::handleDeviceUpdate);
```

fwolter replied:
Did you see the syntactical sugar in my suggestion?

digitaldan replied:
I did not, thanks for the hint

fwolter commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -0,0 +1,352 @@
+                devices.items.forEach(device -> {
+                    ThingTypeUID thingTypeUID = new ThingTypeUID(BINDING_ID, device.deviceFamily);
+                    if (SUPPORTED_DISCOVERY_THING_TYPES_UIDS.contains(thingTypeUID)) {
+                        for (Thing thing : getThing().getThings()) {
+                            ThingHandler handler = thing.getHandler();
+                            if (handler != null && ((MyQDeviceHandler) handler).getSerialNumber()
+                                    .equalsIgnoreCase(device.serialNumber)) {
+                                ((MyQDeviceHandler) handler).handleDeviceUpdate(device);
+                            }
+                        }
```
This seems the same code as in line 126. Can you make it re-useable?

digitaldan replied:
it looks similar, but different enough that i prefer to keep it the way it is

fwolter reviewed (changes requested):
Looks way better with the `ThingHandlerService`!

digitaldan commented:
Thanks, changes done.

digitaldan commented:
Unless there are any major issues, i would love to see this merged soon.  Thanks.

fwolter reviewed (approved):
LGTM

digitaldan commented:
👀

cpmeister commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -0,0 +1,349 @@
+    private static String randomString(int length) {
+        int low = 97; // a-z
+        int high = 122; // A-Z
+        StringBuffer sb = new StringBuffer(length);
```
```suggestion
        StringBuilder sb = new StringBuilder(length);
```

cpmeister commented on bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java:
```diff
@@ -0,0 +1,349 @@
+    private void getAccount() throws InterruptedException {
+        HttpResult result = sendRequest(BASE_URL + "/v5/My?expand=account", HttpMethod.GET, securityToken, null, null);
+        if (HttpStatus.isSuccess(result.responseCode)) {
+            account = gsonUpperCase.fromJson(result.content, AccountDTO.class);
```
You should catch the JsonSyntaxException that is thrown here.

digitaldan replied:
Done.

cpmeister reviewed (changes requested):
Only a few small suggestions.

digitaldan commented:
Just realized my last commit was missing a sign off.  Should be ready now @cpmeister

digitaldan commented:
👀

digitaldan commented:
👀  😢

cpmeister reviewed (approved):
LGTM

morph166955 commented:
Updated to S2225 which includes this.  Everything looks good.  Thank you @digitaldan !!!!!
